### ISSUE TITLE
[strict-boolean-expressions] allow objects as boolean expressions

### ISSUE THREAD

**Issue by erik-beus** [typescript-eslint/typescript-eslint#2916]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": [
      "warn",
      {
        allowString: false,
        allowNumber: false,
        allowNullableObject: true,
        allowNullableBoolean: true,
      },
    ]
  }
}
```

```TS
const test = [
  {
    key: "a",
  },
  { key: "b" },
];
const myVal = test[2];
if (myVal) {
  console.log("success");
}
```

**Expected Result**

I would expect it to be possible to configure the rule to allow objects as boolean expressions - just like it's possible to allow strings/numbers. So something like: `allowObject: true`

**Actual Result**

It's currently not supported, and shows a warning when comparing on `myVal`:

"Unexpected object value in conditional. The condition is always true"

**Additional Info**

In Typescript, whenever an Array is defined as every element being defined, the inferred type will always be defined when accessing an element although it could very well be undefined if there was no value for a certain key (as in the example above).

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.12.0` |
| `@typescript-eslint/parser`        | `4.12.0` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.17.0` |
| `node`                             | `12.16.2` |

**Comment by bradzacher** [comment]
This is a problem with your types - not with the rule.

Consider using the [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess) tsconfig option which make your code safe and provide correct types.

![image](https://user-images.githubusercontent.com/7462525/103574577-0ebfcd80-4e85-11eb-814f-4a098808c3a8.png)
[repl](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBFCm0YF4YG0BQMYG8vZgGt4BPALhgCIBDSgGnwF8HscjSLKAjSmZjALoBuDKEiwAtiQBq1ADYo4iKGgBMwjJs0BLAGYwAFFNlyAlLnxiIIOfAB0ckAHMDlCAFdgwRBEqmRjEA)
